### ISSUE TITLE
Provide a TypeScript type definition

### ISSUE THREAD

**Issue by Zaid-Al-Omari** [versatica/mediasoup#162]
Will you provide a TypeScript definition for those of us that like to work in TypeScript ?

https://microsoft.github.io/TypeSearch/

**Comment by ibc** [comment]
We don't know TypeScript. I'm not against it but we just don't use it so cannot convert the whole project to it.

**Comment by Flamenco** [comment]
You don't have to convert the whole project.  You can start with basic definitions and then slowly migrate to describing your types.  Typescript lets you define an 'overlay' that _describes your API_.  The type definitions do not even have need live inside your project, so you can continue to develop 'typeless' in plain JS.

Not only would this make managing your own project easier, but your users will get IDE completion and type checking for free.

Furthermore, it is the user that decides the degree of type checking in their projects, not the MediaSoup project.  So the developer can not use types at all, fully, or only to a limited degree.

Here is a semi-complete example I am using.  Notice the use of `any` for types I have not defined yet.
Please consider reopening this issue and using this sample as a starting point.
```
namespace MediaSoup {

  export interface Room {
    id: number
    closed: boolean
    rtpCapabilities: any
    peers: Peer[]
  }

  export interface Peer {
    name: string
    closed: boolean
    appData: any
    rtpCapabilities: any
    transports: any[]
    producers: Producer[]
    consumers: Consumer[]
  }

  interface BaseType {
    id: number
    closed: boolean
    appData: any
    rtpParameters: RtpParameters
    kind: 'audio' | 'video' | 'data'
    transport: any
    locallyPaused: boolean
    remotelyPaused: boolean
    paused: boolean
    getStats: Function
  }

  export interface Producer extends BaseType {
    peer: Peer
  }
  
  interface Consumer extends BaseType {
    peer: Peer | undefined // This is undefined when creating an RtpStreamer
    source: Producer //TODO this may be other types
    preferredProfile: any
    effectiveProfile: any
  }

  interface RtpParameters {
    headerExtensions: any[];
    muxId: string
    codecs: Codec[]
  }

  interface Codec {
    rtcpFeedback: any[];
    payloadType: any;
    parameters: any[];
    name: string;
    clockRate: number;
  }

  interface PlainRtpTransport {
    tuple: {
      localIP: string
      localPort: number
      remoteIP: string
      remotePort: number
      protocol: 'udp' | 'tcp'
    }
  }

  export interface RtpStreamer {
    consumer: Consumer
    transport: PlainRtpTransport
  }

  interface ActiveSpeakerDetector {
    close: () => void
  }
}
```

**Comment by ibc** [comment]
HI @Flamenco, I appreciate your explanation. I agree it makes sense and it's not as intrusive as I may consider at the beginning.

However, it would still require that I learn where to create such a "definitions file", how to lint and validate its syntax, which dev-dependencies to install, etc. I don't have time right now to learn about it.

**Comment by Flamenco** [comment]
I will gladly work on it and and add a mediasoup-types repo :)  The API is not very large, and your docs are VERY clear.  Since I posted this, I have already added about 40% of the MediaSoup API.  Even if you are not writing Typescript, IntelliJ adds autocomplete and syntax highlighting if you simply include the types.

Just a suggestion, but the time spent learning and using Typescript will payoff 100-fold.

I have gotten heavy into the soup these days and cannot work without type support.  I wrote an RTSP media soup wrapper with a plugin API, and then a nifty plugin for it with a recording module.  The soup client can publish a stream and then broadcast, record locally, or record on the server.  It works in all browsers and devices, including iOS, with the only caveat that mobile Safari only records audio (not video) locally.  (Safari can still record video on the server however).  I still have some sync issues, but I should have this in production soon.  A single streamset can even be broken up on the server into individual tracks on the fly by sending commands through the web socket.
